src/arch/x86/kernel/boot_sys.c

```diff
@@ -63,23 +63,25 @@ BOOT_CODE static paddr_t find_load_paddr(paddr_t min_paddr, word_t image_size)
     int i;
 
     for (i = 0; i < boot_state.mem_p_regs.count; i++) {
-        paddr_t start = MAX(min_paddr, boot_state.mem_p_regs.list[i].start);
+        paddr_t start = MAX(boot_state.ki_p_reg.end, MAX(min_paddr, boot_state.mem_p_regs.list[i].start));
         paddr_t end = boot_state.mem_p_regs.list[i].end;
         word_t region_size = end - start;
 
-        if (region_size >= image_size) {
+        if (end > start && region_size >= image_size) {
             return start;
         }
     }
 
     return 0;
 }
 
-BOOT_CODE static paddr_t load_boot_module(word_t boot_module_start, paddr_t load_paddr)
+BOOT_CODE static paddr_t load_boot_module(word_t boot_module_start, paddr_t *load_paddr_p)
```

Comment:
If there are added side effects like changing the content of `load_paddr_p` because things are moved, these should be documented in a comment.